src/actions/exams.ts

```diff
@@ -5,28 +5,202 @@ import db from '@/lib/db'
 import { revalidatePath } from 'next/cache'
 import { cache } from 'react'
 
-export const getExams = cache(async () => {
-  const response = await db.exam.findMany({
-    select: {
-      id: true,
-      title: true,
-      description: true,
-      duration: true,
-      price: true,
-    },
-  })
-  return response
-})
+type CreateExamInput = {
+  title: string
+  description: string
+  duration: number
+  price: number
+  numQuestions: number
+}
+
+type UpdateExamInput = {
+  id: string
+  title?: string
+  description?: string
+  duration?: number
+  price?: number
+  numQuestions?: number
+}
 
 interface SubmitExamParams {
   examId: string
   answers: Record<string, number>
   timeSpent: number
   warningCount: number
+  questions: string[]
+}
+
+export const createExam = cache(async (examData: CreateExamInput) => {
+  try {
+    const createdExam = await db.exam.create({
+      data: examData,
+    })
+
+    const responseData = {
+      id: createdExam.id,
+      title: createdExam.title,
+      description: createdExam.description,
+      duration: createdExam.duration,
+      price: createdExam.price,
+      numQuestions: createdExam.numQuestions,
+    }
+
+    return { succes: true, data: responseData }
+  } catch (error) {
+    console.error('Error creating exam:', error)
+    throw new Error('Failed to create exam')
+  }
+})
+
+export const updateExam = cache(async (examData: UpdateExamInput) => {
+  try {
+    const { id, ...updateFields } = examData
+
+    const totalAvailableQuestions = await db.question.findMany()
+
+    if (
+      updateFields.numQuestions !== undefined &&
+      updateFields.numQuestions > totalAvailableQuestions.length
+    ) {
+      throw new Error(
+        'The number of questions are not available. Try fewer questions'
+      )
+    }
+
+    const updatedExam = await db.exam.update({
+      where: { id },
+      data: updateFields,
+    })
+
+    return { success: true, data: updatedExam }
+  } catch (error) {
+    console.error('Error updating exam:', error) // Add more detailed error logging for debugging
+    throw new Error('Failed to update exam')
+  }
+})
+
+export const getExams = async () => {
+  try {
+    const response = await db.exam.findMany({
```

Comment:
Check logic if user can see exam without login we can ignore this else add auth

Reply:
It is a protected route currently which makes more sense, reason being when a user without login redirected to the available exams page (which is a dashboard), it contains all user details, exams taken and everything else on its sidebar. 

If a user is NOT LOGGED IN CURRENTLY and they are to take an exam they should pick the exam and then click on the "Pay & Take" test afterwards he is supposed to be redirected to -> login page -> dashboard ->  available exams ->  select exams that they wish to attempt. Which is most likely a not an ideal UX for an exam portal.